src/utils/floating-ui.ts

```diff
@@ -0,0 +1,392 @@
+   *
+   */
+  placement: LogicalPlacement;
+
+  /**
+   * Updates the position of the component.
+   */
+  reposition(): Promise<void>;
+}
+
+export const FloatingCSS = {
+  animation: "calcite-floating-ui-anim",
+  animationActive: "calcite-floating-ui-anim--active"
+};
+
+function getMiddleware({
+  placement,
+  disableFlip,
+  flipPlacements,
+  offsetDistance,
+  offsetSkidding,
+  arrowEl,
+  type
+}: {
+  placement: LogicalPlacement;
+  disableFlip?: boolean;
+  flipPlacements?: EffectivePlacement[];
+  offsetDistance?: number;
+  offsetSkidding?: number;
+  arrowEl?: HTMLElement;
+  type: UIType;
+}): Middleware[] {
+  const defaultMiddleware = [shift(), hide()];
+
+  if (type === "menu") {
+    return [
+      ...defaultMiddleware,
+      flip({ fallbackPlacements: ["top-start", "top", "top-end", "bottom-start", "bottom", "bottom-end"] })
+    ];
+  }
+
+  if (type === "popover" || type === "tooltip") {
+    const middleware: Middleware[] = [
+      ...defaultMiddleware,
+      offset({
+        mainAxis: typeof offsetDistance === "number" ? offsetDistance : 0,
+        crossAxis: typeof offsetSkidding === "number" ? offsetSkidding : 0
+      })
+    ];
+
+    if (placement === "auto" || placement === "auto-start" || placement === "auto-end") {
+      middleware.push(
+        autoPlacement({ alignment: placement === "auto-start" ? "start" : placement === "auto-end" ? "end" : null })
+      );
+    } else if (!disableFlip) {
+      middleware.push(flip(flipPlacements ? { fallbackPlacements: flipPlacements } : {}));
+    }
+
+    if (arrowEl) {
+      middleware.push(
+        arrow({
+          element: arrowEl
+        })
+      );
+    }
+
+    return middleware;
+  }
+
+  return [];
+}
+
+export function filterComputedPlacements(placements: string[], el: HTMLElement): EffectivePlacement[] {
+  const filteredPlacements = placements.filter((placement: EffectivePlacement) =>
+    effectivePlacements.includes(placement)
+  ) as EffectivePlacement[];
+
+  if (filteredPlacements.length !== placements.length) {
+    console.warn(
+      `${el.tagName}: Invalid value found in: flipPlacements. Try any of these: ${effectivePlacements
+        .map((placement) => `"${placement}"`)
+        .join(", ")
+        .trim()}`,
+      { el }
+    );
+  }
+
+  return filteredPlacements;
+}
+
+/*
+In floating-ui, "*-start" and "*-end" are already flipped in RTL.
+There is no need for our "*-leading" and "*-trailing" values anymore.
+https://github.com/floating-ui/floating-ui/issues/1530
+https://github.com/floating-ui/floating-ui/issues/1563
+*/
+export function getEffectivePlacement(floatingEl: HTMLElement, placement: LogicalPlacement): EffectivePlacement {
+  const placements = ["left", "right"];
+
+  if (getElementDir(floatingEl) === "rtl") {
+    placements.reverse();
+  }
+
+  return placement
+    .replace(/-leading/gi, "-start") // todo: remove. Has already been removed in typings.
```

Comment:
Suggestion: to avoid unnecessary TS errors, you can add the deprecated values in a `DeprecatedPlacement` type or similar.